File: internal/apps/manifest.go
package apps

import (
	"encoding/json"
	"fmt"
	"os"
	"path/filepath"
	"strings"
)

// Known capability types that apps can provide.
const (
	CapGateway  = "gateway"
	CapVision   = "vision"
	CapBrowser  = "browser"
	CapComm     = "comm"
	CapUI       = "ui"
	CapSchedule = "schedule"
)

// Capability prefixes for parameterized capabilities.
const (
	CapPrefixTool    = "tool:"
	CapPrefixChannel = "channel:"
)

// Known permission prefixes.
const (
	// Storage & Config
	PermPrefixNetwork    = "network:"
	PermPrefixFilesystem = "filesystem:"
	PermPrefixSettings   = "settings:"
	PermPrefixCapability = "capability:"

	// Agent Core
	PermPrefixMemory  = "memory:"
	PermPrefixSession = "session:"
	PermPrefixContext = "context:"

	// Execution
	PermPrefixTool     = "tool:"
	PermPrefixShell    = "shell:"
	PermPrefixSubagent = "subagent:"
	PermPrefixLane     = "lane:"

	// Communication
	PermPrefixChannel      = "channel:"
	PermPrefixComm         = "comm:"
	PermPrefixNotification = "notification:"

	// Knowledge
	PermPrefixEmbedding = "embedding:"
	PermPrefixSkill     = "skill:"
	PermPrefixAdvisor   = "advisor:"

	// AI
	PermPrefixModel = "model:"
	PermPrefixMCP   = "mcp:"

	// Storage
	PermPrefixDatabase = "database:"
	PermPrefixStorage  = "storage:"

	// System
	PermPrefixSchedule = "schedule:"
	PermPrefixVoice    = "voice:"
	PermPrefixBrowser  = "browser:"
	PermPrefixOAuth    = "oauth:"
	PermPrefixUser     = "user:"
)

// AppManifest represents an app's manifest.json — the "plist" for Nebo apps.
// Apps declare what they provide (capabilities) and what they need (permissions).
type AppManifest struct {
	ID          string            `json:"id"`
	Name        string            `json:"name"`
	Version     string            `json:"version"`
	Description string            `json:"description,omitempty"`
	Runtime     string            `json:"runtime"`  // "local" or "remote"
	Protocol    string            `json:"protocol"` // "grpc"
	Signature      ManifestSignature `json:"signature,omitempty"`
	StartupTimeout int               `json:"startup_timeout,omitempty"` // seconds, 0 = default (10s)
	Provides       []string          `json:"provides"`
	Permissions []string          `json:"permissions"`
	OAuth       []OAuthRequirement `json:"oauth,omitempty"`
}

// OAuthRequirement declares an OAuth provider and scopes the app needs.
// Nebo's OAuth broker handles the entire flow — apps receive tokens automatically.
type OAuthRequirement struct {
	Provider string   `json:"provider"` // "google", "microsoft", "github"
	Scopes   []string `json:"scopes"`   // e.g. ["https://www.googleapis.com/auth/calendar"]
}

// ManifestSignature holds NeboLoop code-signing information.
type ManifestSignature struct {
	Algorithm string `json:"algorithm,omitempty"` // e.g. "ed25519"
	PublicKey string `json:"public_key,omitempty"`
	Signature string `json:"signature,omitempty"` // Base64 signature of manifest (minus signature field)
	BinarySig string `json:"binary_sig,omitempty"` // Base64 signature of binary
}


// LoadManifest reads and validates a manifest.json from an app directory.
func LoadManifest(dir string) (*AppManifest, error) {
	path := filepath.Join(dir, "manifest.json")
	data, err := os.ReadFile(path)
	if err != nil {
		return nil, fmt.Errorf("read manifest: %w", err)
	}

	var m AppManifest
	if err := json.Unmarshal(data, &m); err != nil {
		return nil, fmt.Errorf("parse manifest: %w", err)
	}

	if err := ValidateManifest(&m); err != nil {
		return nil, fmt.Errorf("invalid manifest: %w", err)
	}

	return &m, nil
}

// ValidateManifest checks that required fields are present and values are valid.
func ValidateManifest(m *AppManifest) error {
	if m.ID == "" {
		return fmt.Errorf("missing required field: id")
	}
	if m.Name == "" {
		return fmt.Errorf("missing required field: name")
	}
	if m.Version == "" {
		return fmt.Errorf("missing required field: version")
	}
	if len(m.Provides) == 0 {
		return fmt.Errorf("missing required field: provides (must declare at least one capability)")
	}
	if m.Protocol != "" && m.Protocol != "grpc" {
		return fmt.Errorf("unsupported protocol: %s (only grpc is supported)", m.Protocol)
	}
	if m.Runtime != "" && m.Runtime != "local" && m.Runtime != "remote" {
		return fmt.Errorf("unsupported runtime: %s (must be local or remote)", m.Runtime)
	}
	if m.StartupTimeout < 0 || m.StartupTimeout > 120 {
		return fmt.Errorf("startup_timeout must be between 0 and 120 seconds (got %d)", m.StartupTimeout)
	}

	for _, cap := range m.Provides {
		if !isValidCapability(cap) {
			return fmt.Errorf("invalid capability: %s", cap)
		}
	}

	for _, perm := range m.Permissions {
		if !isValidPermission(perm) {
			return fmt.Errorf("invalid permission: %s", perm)
		}
	}

	return nil
}

// VerifySignature is now implemented in signing.go as VerifyAppSignatures.
// This stub exists for backward compatibility during the transition.
// When keyProvider is nil (dev mode), signature verification is skipped.
func VerifySignature(m *AppManifest, binaryPath string) error {
	return nil
}

// HasCapability returns true if the manifest declares the given capability.
func HasCapability(m *AppManifest, cap string) bool {
	for _, c := range m.Provides {
		if c == cap {
			return true
		}
	}
	return false
}

// HasCapabilityPrefix returns true if any capability starts with the given prefix.
func HasCapabilityPrefix(m *AppManifest, prefix string) bool {
	for _, c := range m.Provides {
		if strings.HasPrefix(c, prefix) {
			return true
		}
	}
	return false
}

// HasPermissionPrefix returns true if any permission starts with the given prefix.
func HasPermissionPrefix(m *AppManifest, prefix string) bool {
	for _, p := range m.Permissions {
		if strings.HasPrefix(p, prefix) {
			return true
		}
	}
	return false
}

// CheckPermission returns true if the manifest includes the given permission.
func CheckPermission(m *AppManifest, perm string) bool {
	for _, p := range m.Permissions {
		if p == perm {
			return true
		}
		// Check wildcard: "network:*" matches any "network:..." check
		prefix, _, ok := strings.Cut(p, ":")
		if ok {
			reqPrefix, _, reqOk := strings.Cut(perm, ":")
			if reqOk && prefix == reqPrefix && strings.HasSuffix(p, ":*") {
				return true
			}
		}
	}
	return false
}

func isValidCapability(cap string) bool {
	switch cap {
	case CapGateway, CapVision, CapBrowser, CapComm, CapUI, CapSchedule:
		return true
	}
	if strings.HasPrefix(cap, CapPrefixTool) || strings.HasPrefix(cap, CapPrefixChannel) {
		return true
	}
	return false
}

// validPermissionPrefixes lists all recognized permission prefixes.
// Unknown prefixes are rejected — apps must use the documented taxonomy.
var validPermissionPrefixes = []string{
	PermPrefixNetwork, PermPrefixFilesystem, PermPrefixSettings, PermPrefixCapability,
	PermPrefixMemory, PermPrefixSession, PermPrefixContext,
	PermPrefixTool, PermPrefixShell, PermPrefixSubagent, PermPrefixLane,
	PermPrefixChannel, PermPrefixComm, PermPrefixNotification,
	PermPrefixEmbedding, PermPrefixSkill, PermPrefixAdvisor,
	PermPrefixModel, PermPrefixMCP,
	PermPrefixDatabase, PermPrefixStorage,
	PermPrefixSchedule, PermPrefixVoice, PermPrefixBrowser, PermPrefixOAuth, PermPrefixUser,
}

// validPermissionSuffixes maps each prefix to its allowed suffixes.
// nil means the prefix accepts flexible identifier-style suffixes (hostnames, provider names).
var validPermissionSuffixes = map[string][]string{
	// Storage & Config
	PermPrefixNetwork:    nil, // flexible: "outbound", host:port patterns, "*"
	PermPrefixFilesystem: {"read", "write"},
	PermPrefixSettings:   {"read", "write"},
	PermPrefixCapability: {"register"},

	// Agent Core
	PermPrefixMemory:  {"read", "write"},
	PermPrefixSession: {"read", "write", "create"},
	PermPrefixContext: {"read"},

	// Execution
	PermPrefixTool:     {"file", "shell", "web", "agent", "skill"},
	PermPrefixShell:    {"exec"},
	PermPrefixSubagent: {"spawn"},
	PermPrefixLane:     {"enqueue"},

	// Communication
	PermPrefixChannel:      {"send", "receive"},
	PermPrefixComm:         {"send", "receive"},
	PermPrefixNotification: {"send"},

	// Knowledge
	PermPrefixEmbedding: {"search", "store"},
	PermPrefixSkill:     {"invoke"},
	PermPrefixAdvisor:   {"consult"},

	// AI
	PermPrefixModel: {"chat", "embed"},
	PermPrefixMCP:   {"connect"},

	// Storage
	PermPrefixDatabase: {"query", "read", "write"},
	PermPrefixStorage:  {"read", "write"},

	// System
	PermPrefixSchedule: {"create", "delete", "list"},
	PermPrefixVoice:    {"record"},
	PermPrefixBrowser:  {"navigate"},
	PermPrefixOAuth:    nil, // flexible: provider names ("google", "microsoft", etc.)
	PermPrefixUser:     {"token", "id"},
}

func isValidPermission(perm string) bool {
	for _, prefix := range validPermissionPrefixes {
		if !strings.HasPrefix(perm, prefix) {
			continue
		}
		suffix := perm[len(prefix):]
		if suffix == "" {
			return false // bare prefix with no value
		}
		if suffix == "*" {
			return true // wildcard always valid
		}
		allowed := validPermissionSuffixes[prefix]
		if allowed == nil {
			// Flexible prefix — validate format only
			return isValidPermissionIdentifier(suffix)
		}
		for _, v := range allowed {
			if suffix == v {
				return true
			}
		}
		return false
	}
	return false
}

// isValidPermissionIdentifier checks that a flexible suffix contains only
// lowercase alphanumeric characters, dots, hyphens, colons, and underscores.
// This covers hostnames (api.example.com), ports (host:443), and provider names (google).
func isValidPermissionIdentifier(s string) bool {
	for _, c := range s {
		if !((c >= 'a' && c <= 'z') || (c >= '0' && c <= '9') || c == '.' || c == '-' || c == ':' || c == '_') {
			return false
		}
	}
	return true
}

File: crates/napp/src/manifest.rs
use serde::{Deserialize, Serialize};

use crate::NappError;

/// Parsed qualified name: `@org/type/name`.
///
/// Valid types: `skills`, `workflows`, `roles`.
#[derive(Debug, Clone, PartialEq, Eq)]
pub struct QualifiedName {
    pub org: String,
    pub artifact_type: String,
    pub artifact_name: String,
}

impl QualifiedName {
    /// Parse `@org/type/name` format.
    pub fn parse(name: &str) -> Result<Self, NappError> {
        let s = name.strip_prefix('@').ok_or_else(|| {
            NappError::Manifest(format!("qualified name must start with '@': {}", name))
        })?;

        let parts: Vec<&str> = s.splitn(3, '/').collect();
        if parts.len() != 3 {
            return Err(NappError::Manifest(format!(
                "qualified name must be @org/type/name: {}",
                name
            )));
        }

        let artifact_type = parts[1];
        if !["skills", "workflows", "roles"].contains(&artifact_type) {
            return Err(NappError::Manifest(format!(
                "invalid artifact type '{}' in qualified name (expected skills/workflows/roles)",
                artifact_type
            )));
        }

        Ok(Self {
            org: parts[0].to_string(),
            artifact_type: artifact_type.to_string(),
            artifact_name: parts[2].to_string(),
        })
    }

    /// Format as the full qualified string.
    pub fn to_string(&self) -> String {
        format!("@{}/{}/{}", self.org, self.artifact_type, self.artifact_name)
    }
}

/// Package manifest (manifest.json) — universal envelope for all artifact types.
///
/// Every artifact (skill, tool, workflow, role) includes a manifest.json with
/// identity fields (id, name, version, type, description). Tool-specific fields
/// (provides, permissions, implements, etc.) default to empty for non-tool artifacts.
#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct Manifest {
    pub id: String,
    pub name: String,
    pub version: String,
    /// Artifact type: "skill", "tool", "workflow", or "role".
    #[serde(rename = "type", default)]
    pub artifact_type: String,
    #[serde(default)]
    pub description: String,
    /// Publisher name.
    #[serde(default)]
    pub author: String,
    /// Marketplace code (assigned on publish).
    #[serde(default)]
    pub code: String,
    /// Categorization tags.
    #[serde(default)]
    pub tags: Vec<String>,
    // -- Tool-specific fields (ignored for non-tool artifacts) --
    #[serde(default = "default_runtime")]
    pub runtime: String,
    #[serde(default = "default_protocol")]
    pub protocol: String,
    #[serde(default)]
    pub signature: Option<ManifestSignature>,
    #[serde(default)]
    pub startup_timeout: u32,
    #[serde(default)]
    pub provides: Vec<String>,
    #[serde(default)]
    pub permissions: Vec<String>,
    #[serde(default)]
    pub overrides: Vec<String>,
    #[serde(default)]
    pub oauth: Vec<OAuthRequirement>,
    #[serde(default)]
    pub implements: Vec<String>,
}

fn default_runtime() -> String {
    "local".to_string()
}
fn default_protocol() -> String {
    "grpc".to_string()
}

/// Code signing info in the manifest.
#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct ManifestSignature {
    #[serde(default)]
    pub algorithm: String,
    #[serde(default)]
    pub key_id: String,
    /// SHA256 hash of the binary — verified on every launch.
    #[serde(default)]
    pub binary_hash: String,
    /// Signature over the manifest content.
    #[serde(default)]
    pub manifest_signature: String,
    /// Signature over the binary hash.
    #[serde(default)]
    pub binary_signature: String,
}

/// OAuth requirement for a tool.
#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct OAuthRequirement {
    pub provider: String,
    #[serde(default)]
    pub scopes: Vec<String>,
}

/// Valid capabilities a tool can provide.
const VALID_CAPABILITIES: &[&str] = &[
    "gateway", "vision", "browser", "comm", "ui", "schedule", "hooks",
];

/// Valid permission prefixes.
const VALID_PERMISSION_PREFIXES: &[&str] = &[
    "network:", "filesystem:", "settings:", "capability:", "memory:",
    "session:", "context:", "tool:", "shell:", "subagent:", "lane:",
    "channel:", "comm:", "notification:", "embedding:", "skill:",
    "advisor:", "model:", "mcp:", "database:", "storage:", "schedule:",
    "voice:", "browser:", "oauth:", "user:", "hook:",
];

impl Manifest {
    /// Load manifest from a JSON file.
    pub fn load(path: &std::path::Path) -> Result<Self, NappError> {
        let data = std::fs::read_to_string(path)?;
        let manifest: Self = serde_json::from_str(&data)?;
        Ok(manifest)
    }

    /// Backward-compatible ID accessor.
    ///
    /// If `name` is a qualified name (`@org/type/name`), returns `&self.name`.
    /// Otherwise returns `&self.id`.
    pub fn id(&self) -> &str {
        if self.name.starts_with('@') {
            &self.name
        } else {
            &self.id
        }
    }

    /// Parse the qualified name if `name` starts with `@`.
    pub fn qualified_name(&self) -> Option<QualifiedName> {
        if self.name.starts_with('@') {
            QualifiedName::parse(&self.name).ok()
        } else {
            None
        }
    }

    /// Validate the manifest.
    pub fn validate(&self) -> Result<(), NappError> {
        if self.id.is_empty() && !self.name.starts_with('@') {
            return Err(NappError::Manifest("id is required".into()));
        }
        if self.name.is_empty() {
            return Err(NappError::Manifest("name is required".into()));
        }
        if self.version.is_empty() {
            return Err(NappError::Manifest("version is required".into()));
        }

        // Validate qualified name format when name starts with @
        if self.name.starts_with('@') {
            QualifiedName::parse(&self.name)?;
        }

        // Validate capabilities
        for cap in &self.provides {
            let base = cap.split(':').next().unwrap_or(cap);
            if base != "tool" && base != "channel" && !VALID_CAPABILITIES.contains(&base) {
                return Err(NappError::Manifest(format!(
                    "unknown capability: {}",
                    cap
                )));
            }
        }

        // Validate permissions
        for perm in &self.permissions {
            let valid = VALID_PERMISSION_PREFIXES
                .iter()
                .any(|prefix| perm.starts_with(prefix));
            if !valid {
                return Err(NappError::Manifest(format!(
                    "unknown permission: {}",
                    perm
                )));
            }
        }

        // Overrides require hook: permission
        for override_name in &self.overrides {
            let required_perm = format!("hook:{}", override_name);
            if !self.permissions.contains(&required_perm) {
                return Err(NappError::Manifest(format!(
                    "override '{}' requires permission '{}'",
                    override_name, required_perm
                )));
            }
        }

        // Startup timeout cap
        if self.startup_timeout > 120 {
            return Err(NappError::Manifest(
                "startup_timeout must be <= 120 seconds".into(),
            ));
        }

        Ok(())
    }

    /// Check if the tool has a specific permission.
    pub fn has_permission(&self, perm: &str) -> bool {
        self.permissions.iter().any(|p| {
            p == perm || (p.ends_with(':') && perm.starts_with(p)) || p == &format!("{}:*", perm.split(':').next().unwrap_or(""))
        })
    }

    /// Check if the tool has a permission with a prefix (wildcard support).
    pub fn has_permission_prefix(&self, prefix: &str) -> bool {
        self.permissions.iter().any(|p| p.starts_with(prefix))
    }

    /// Get the effective startup timeout (default 10s).
    pub fn effective_startup_timeout(&self) -> u32 {
        if self.startup_timeout == 0 {
            10
        } else {
            self.startup_timeout.min(120)
        }
    }
}

#[cfg(test)]
mod tests {
    use super::*;

    #[test]
    fn test_validate_valid_manifest() {
        let m = Manifest {
            id: "test-tool".into(),
            name: "Test Tool".into(),
            version: "1.0.0".into(),
            description: "A test tool".into(),
            startup_timeout: 10,
            provides: vec!["gateway".into(), "tool:search".into()],
            permissions: vec!["network:*".into(), "tool:web".into()],
            ..Default::default()
        };
        assert!(m.validate().is_ok());
    }

    #[test]
    fn test_validate_missing_id() {
        let m = Manifest {
            id: "".into(),
            name: "Test".into(),
            version: "1.0.0".into(),
            ..Default::default()
        };
        assert!(m.validate().is_err());
    }

    #[test]
    fn test_has_permission() {
        let m = Manifest {
            id: "x".into(),
            name: "X".into(),
            version: "1".into(),
            permissions: vec!["network:*".into(), "tool:web".into()],
            ..Default::default()
        };
        assert!(m.has_permission("tool:web"));
        assert!(m.has_permission("network:example.com"));
    }

    #[test]
    fn test_startup_timeout() {
        let mut m = Manifest::default();
        m.id = "x".into();
        m.name = "X".into();
        m.version = "1".into();
        assert_eq!(m.effective_startup_timeout(), 10);
        m.startup_timeout = 60;
        assert_eq!(m.effective_startup_timeout(), 60);
        m.startup_timeout = 200;
        assert_eq!(m.effective_startup_timeout(), 120);
    }

    #[test]
    fn test_qualified_name_parse() {
        let qn = QualifiedName::parse("@acme/skills/crm-lookup").unwrap();
        assert_eq!(qn.org, "acme");
        assert_eq!(qn.artifact_type, "skills");
        assert_eq!(qn.artifact_name, "crm-lookup");
    }

    #[test]
    fn test_qualified_name_invalid() {
        assert!(QualifiedName::parse("not-qualified").is_err());
        assert!(QualifiedName::parse("@acme/invalid_type/name").is_err());
        assert!(QualifiedName::parse("@acme/skills").is_err());
        assert!(QualifiedName::parse("@acme/tools/name").is_err()); // tools no longer valid
    }

    #[test]
    fn test_manifest_id_accessor() {
        let mut m = Manifest::default();
        m.id = "legacy-id".into();
        m.name = "Legacy Name".into();
        m.version = "1.0".into();
        assert_eq!(m.id(), "legacy-id");

        m.name = "@acme/skills/crm-lookup".into();
        assert_eq!(m.id(), "@acme/skills/crm-lookup");
    }

    #[test]
    fn test_implements_field() {
        let json = r#"{
            "id": "crm-tool",
            "name": "CRM Tool",
            "version": "1.0.0",
            "implements": ["crm-lookup", "contact-search"]
        }"#;
        let m: Manifest = serde_json::from_str(json).unwrap();
        assert_eq!(m.implements, vec!["crm-lookup", "contact-search"]);
    }
}

impl Default for Manifest {
    fn default() -> Self {
        Self {
            id: String::new(),
            name: String::new(),
            version: String::new(),
            artifact_type: String::new(),
            description: String::new(),
            author: String::new(),
            code: String::new(),
            tags: vec![],
            runtime: "local".into(),
            protocol: "grpc".into(),
            signature: None,
            startup_timeout: 0,
            provides: vec![],
            permissions: vec![],
            overrides: vec![],
            oauth: vec![],
            implements: vec![],
        }
    }
}
